--- vimprobable2/main.c	2010-12-11 19:54:52.000000000 +0100
+++ vimprobable2atomuri/main.c	2011-02-06 19:36:31.173332973 +0100
@@ -9,6 +9,8 @@
 */
 
 #include <X11/Xlib.h>
+#include <X11/Xatom.h>
+#include <gdk/gdkx.h>
 #include "includes.h"
 #include "vimprobable.h"
 #include "utilities.h"
@@ -126,6 +128,7 @@
 static GdkNativeWindow embed = 0;
 static char *configfile = NULL;
 static char *winid = NULL;
+static Atom atom_uri = NULL;
 
 static char rememberedURI[128] = "";
 static char inputKey[5];
@@ -1908,6 +1911,8 @@
     gdk_color_parse(ssl ? sslcolor : statuscolor, &color);
     gtk_widget_modify_fg(GTK_WIDGET(status_url), GTK_STATE_NORMAL, &color);
     gtk_widget_modify_fg(GTK_WIDGET(status_state), GTK_STATE_NORMAL, &color);
+    XSync(GDK_DISPLAY(), False);
+    XChangeProperty(GDK_DISPLAY(), GDK_WINDOW_XID(GTK_WIDGET(window)->window), atom_uri, XA_STRING, 8, PropModeReplace, (unsigned char *)uri, strlen(uri) + 1);
 }
 
 void
@@ -2025,6 +2030,7 @@
     gtk_widget_modify_bg(eventbox, GTK_STATE_NORMAL, &bg);
     gtk_widget_set_name(GTK_WIDGET(window), "Vimprobable2");
     gtk_window_set_geometry_hints(window, NULL, &hints, GDK_HINT_MIN_SIZE);
+    atom_uri = XInternAtom(GDK_DISPLAY(),"_URI", False);
 
 #ifdef DISABLE_SCROLLBAR
     viewport = gtk_scrolled_window_new(NULL, NULL);
