--- src/zathura/zathura.c	2011-04-04 12:07:28.553333479 +0200
+++ src/zathuraatomuri2/zathura.c	2011-04-04 15:17:06.130000386 +0200
@@ -18,6 +18,9 @@
 #include <gtk/gtk.h>
 #include <gdk/gdkkeysyms.h>
 
+#include <gdk/gdkx.h>
+#include <X11/Xatom.h>
+
 /* macros */
 #define LENGTH(x) (sizeof(x)/sizeof((x)[0]))
 #define CLEAN(m) (m & ~(GDK_MOD2_MASK) & ~(GDK_BUTTON1_MASK) & ~(GDK_BUTTON2_MASK) & ~(GDK_BUTTON3_MASK) & ~(GDK_BUTTON4_MASK) & ~(GDK_BUTTON5_MASK) & ~(GDK_LEAVE_NOTIFY_MASK))
@@ -198,6 +201,8 @@
   int page;
 } Bookmark;
 
+Atom atom_uri = NULL;
+
 /* zathura */
 struct
 {
@@ -714,6 +719,8 @@
   Zathura.UI.inputbar          = GTK_ENTRY(gtk_entry_new());
   Zathura.UI.document            = gtk_event_box_new();
 
+  atom_uri = XInternAtom(GDK_DISPLAY(),"_URI", False);
+
   /* window */
   gtk_window_set_title(GTK_WINDOW(Zathura.UI.window), "zathura");
   GdkGeometry hints = { 1, 1 };
@@ -1425,6 +1432,19 @@
   set_page(start_page);
   update_status();
 
+  if (Zathura.UI.window->window)
+  {
+    XSync(GDK_DISPLAY(), False);
+    XChangeProperty(
+            GDK_DISPLAY(),
+            GDK_WINDOW_XID(Zathura.UI.window->window),
+            atom_uri,
+            XA_STRING,
+            8,
+            PropModeReplace,
+            (unsigned char *)file,strlen(file) + 1);
+  }
+
   g_static_mutex_unlock(&(Zathura.Lock.pdf_obj_lock));
   isc_abort(NULL);
   return TRUE;
@@ -4723,6 +4743,15 @@
   init_bookmarks();
   init_look();
 
+  gtk_widget_show_all(GTK_WIDGET(Zathura.UI.window));
+  gtk_widget_grab_focus(GTK_WIDGET(Zathura.UI.view));
+
+  if(!Zathura.Global.show_inputbar)
+    gtk_widget_hide(GTK_WIDGET(Zathura.UI.inputbar));
+
+  if(!Zathura.Global.show_statusbar)
+    gtk_widget_hide(GTK_WIDGET(Zathura.UI.statusbar));
+
   if(argc > 1)
   {
     char* password = (argc == 3) ? argv[2] : NULL;
@@ -4735,15 +4764,6 @@
   switch_view(Zathura.UI.document);
   update_status();
 
-  gtk_widget_show_all(GTK_WIDGET(Zathura.UI.window));
-  gtk_widget_grab_focus(GTK_WIDGET(Zathura.UI.view));
-
-  if(!Zathura.Global.show_inputbar)
-    gtk_widget_hide(GTK_WIDGET(Zathura.UI.inputbar));
-
-  if(!Zathura.Global.show_statusbar)
-    gtk_widget_hide(GTK_WIDGET(Zathura.UI.statusbar));
-
   gdk_threads_enter();
   gtk_main();
   gdk_threads_leave();
